README
D64::Disk::Dir Version 0.02
===========================

D64::File::Dir is a Perl module providing an abstract layer above
D64::Disk::Image module, enabling user to handle D64 disk image directories
in a higher-level object-oriented way.

PREREQUISITES

    Array::Iterator >= 0.06
    D64::Disk::Image >= 0.01

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

COPYRIGHT AND LICENCE

This module is licensed under a slightly modified BSD license, the same terms
as Per Olofsson's "diskimage.c" library and D64::Disk::Image Perl package it is
based on, license contents are repeated below.

Copyright (C) 2003-2006, Per Olofsson
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

- Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.
- Redistributions in binary form must reproduce the above copyright notice, this
  list of conditions and the following disclaimer in the documentation and/or
  other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

diskimage.c website: http://www.paradroid.net/diskimage/


Changes
Revision history for Perl extension D64::Disk::Dir.

0.02  2011-01-13
    - fixed a typo in the NAME section of POD for D64::Disk::Dir::Iterator

0.01  2011-01-09
    - initial version (supports reading D64/D71/D81 disk image directories,
      and accessing individual directory entries)


lib/D64/Disk/Dir.pm
package D64::Disk::Dir;

=head1 NAME

D64::Disk::Dir - Handling entire Commodore (D64/D71/D81) disk image directories

=head1 SYNOPSIS

  use D64::Disk::Dir;

  # Read entire D64/D71/D81 disk image directory from file on disk in one step:
  my $d64DiskDirObj = D64::Disk::Dir->new($filename);

  # Read entire D64/D71/D81 disk image directory from file on disk in two steps:
  my $d64DiskDirObj = D64::Disk::Dir->new();
  my $readOK = $d64DiskDirObj->read_dir($filename);

  # Read new D64/D71/D81 disk directory replacing currently loaded dir with it:
  my $readOK = $d64DiskDirObj->read_dir($filename);

  # Get disk directory title converted to ASCII string:
  my $convert2ascii = 1;
  my $title = $d64DiskDirObj->get_title($convert2ascii);

  # Get disk directory ID converted to ASCII string:
  my $convert2ascii = 1;
  my $diskID = $d64DiskDirObj->get_id($convert2ascii);

  # Get number of blocks free:
  my $blocksFree = $d64DiskDirObj->get_blocks_free();

  # Get number of directory entries:
  my $num_entries = $d64DiskDirObj->num_entries();

  # Get directory entry at the specified position:
  my $entryObj = $d64DiskDirObj->get_entry($index);

  # Get binary file data from a directory entry at the specified position:
  my $data = $d64DiskDirObj->get_file_data($index);

  # Print out the entire directory content to the standard output:
  $d64DiskDirObj->print_dir();

=head1 DESCRIPTION

This package provides an abstract layer above D64::Disk::Image module, enabling user to handle D64 disk image directories in a higher-level object-oriented way.

=head1 METHODS

=cut

use bytes;
use strict;
use warnings;

use base qw( Exporter );
our %EXPORT_TAGS = ();
$EXPORT_TAGS{'all'} = [];
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw();

our $VERSION = '0.01';

use Carp qw/carp croak verbose/;

use D64::Disk::Dir::Entry;
use D64::Disk::Image qw(:all);

# Mapping file types onto file type constants:
our %file_type_constants = (
    'del' => T_DEL,
    'seq' => T_SEQ,
    'prg' => T_PRG,
    'usr' => T_USR,
    'rel' => T_REL,
    'cbm' => T_CBM,
    'dir' => T_DIR,
    '???' => 0xFF,
);

=head2 new

Create empty C<D64::Disk::Dir> object without loading disk image directory yet:

  my $d64DiskDirObj = D64::Disk::Dir->new();

Create new C<D64::Disk::Dir> object and read entire D64/D71/D81 disk image directory from file on disk for further access.

  my $d64DiskDirObj = D64::Disk::Dir->new($filename);

A valid C<D64::Disk::Dir> object is returned upon success, an undefined value otherwise.

=cut

sub new {
    my $this = shift;
    my $class = ref($this) || $this;
    my $self = {};
    bless $self, $class;
    my $initOK = $self->_initialize(@_);
    if ($initOK) {
        return $self;
    }
    else {
        return undef;
    }
}

sub _initialize {
    my $self = shift;
    my $filename = shift;
    # Read entire disk image directory:
    if (defined $filename) {
        my $readOK = $self->read_dir($filename);
        return 0 unless $readOK;
    }
    return 1;
}

sub _check_dir_read {
    my $self = shift;
    # Raise error if directory has not been read yet:
    croak "Unable to perform requested operation, because disk image directory has not been read yet" if $self->{'DIR_READ'} == 0;
}

sub _init_dir {
    my $self = shift;
    # Directory has not been read yet:
    $self->{'DIR_READ'} = 0;
    $self->_release_d64_image();
    $self->_clear_dir_entries();
    delete $self->{'D64_FILE_NAME'};
    delete $self->{'DIR_INFO'};
}

=head2 read_dir

Read entire D64/D71/D81 disk image directory from file on disk, replacing currently loaded directory (if any).

  $d64DiskDirObj->read_dir($filename);

Returns true value upon success, and false otherwise.

=cut

sub read_dir {
    my $self = shift;
    my $filename = shift;
    # We do not verify file existence here, D64::Disk::Image module croaks on inexisting files:
    $self->_init_dir();
    # Load image into RAM:
    my $d64DiskImageObj = D64::Disk::Image->load_image($filename);
    $self->{'D64_DISK_IMAGE'} = $d64DiskImageObj;
    $self->{'D64_FILE_NAME'} = $filename;
    # Open directory for reading:
    my $dir = $d64DiskImageObj->open('$', T_PRG, F_READ);
    # Get disk-wide directory information:
    $self->_get_dir_info($dir);
    # Read directory blocks:
    my $readOK = $self->_read_dir_blocks($dir);
    return 0 unless $readOK;
    # Close directory:
    $dir->close();
    # Store D64 disk image filename for further checks:
    $self->{'FILENAME'} = $filename;
    # Directory has been read successfully:
    $self->{'DIR_READ'} = 1;
    return 1;
    # There was a problem reading directory:
    return 0;
}

sub _read_dir_blocks {
    my $self = shift;
    my $dir = shift;
    # Read first block into buffer:
    my ($counter, $buffer) = $dir->read(254);
    if ($counter != 254) {
        carp 'BAM read failed';
        return 0;
    }
    # Read directory blocks:
    while (1) {
        my ($counter, $buffer) = $dir->read(254);
        last unless $counter == 254;
        for (my $offset = -2; $offset < 254; $offset += 32) {
            # If file type != 0:
            my $file_type = ord (substr $buffer, $offset + 2, 1);
            if ($file_type != 0) {
                # Create new D64::Disk::Dir::Entry object:
                my $bytes = substr $buffer, $offset + 2, 30;
                my $entryObj = D64::Disk::Dir::Entry->new($bytes);
                unless (defined $entryObj) {
                    carp 'Directory blocks read failed';
                    return 0;
                }
                # Add it to the list of directory entries:
                $self->_add_dir_entry($entryObj);
            }
        }
    }
    return 1;
}

sub _get_dir_info {
    my $self = shift;
    my $dir = shift;
    my $d64DiskImageObj = $self->{'D64_DISK_IMAGE'};
    # Get number of blocks free:
    my $blocksFree = $d64DiskImageObj->blocksfree();
    # Get title and ID:
    my ($title, $id) = $d64DiskImageObj->title();
    $title = D64::Disk::Image->name_from_rawname($title);
    $id = D64::Disk::Image->name_from_rawname($id);
    # Store directory details in a hash:
    $self->{'DIR_INFO'} = {
        'TITLE'       => $title,
        'ID'          => $id,
        'BLOCKS_FREE' => $blocksFree,
    };
}

sub _add_dir_entry {
    my $self = shift;
    my $entryObj = shift;
    push @{$self->{'DIR_ENTRIES'}}, $entryObj;
}

sub _get_dir_entries {
    my $self = shift;
    my $entries = $self->{'DIR_ENTRIES'};
    $entries = [] unless defined $entries;
    return $entries;
}

sub _clear_dir_entries {
    my $self = shift;
    $self->{'DIR_ENTRIES'} = [];
}

=head2 get_title

Get 16 character disk directory title (PETSCII string):

  my $convert2ascii = 0;
  my $title = $d64DiskDirObj->get_title($convert2ascii);

Get disk directory title converted to ASCII string:

  my $convert2ascii = 1;
  my $title = $d64DiskDirObj->get_title($convert2ascii);

=cut

sub get_title {
    my $self = shift;
    my $convert2ascii = shift;
    $self->_check_dir_read();
    my $title = $self->{'DIR_INFO'}->{'TITLE'};
    # Convert title to ASCII when necessary:
    $title = D64::Disk::Image->petscii_to_ascii($title) if $convert2ascii;
    return $title;
}

=head2 get_id

Get 2 character disk directory ID (PETSCII string):

  my $convert2ascii = 0;
  my $diskID = $d64DiskDirObj->get_id($convert2ascii);

Get disk directory ID converted to ASCII string:

  my $convert2ascii = 1;
  my $diskID = $d64DiskDirObj->get_id($convert2ascii);

=cut

sub get_id {
    my $self = shift;
    my $convert2ascii = shift;
    $self->_check_dir_read();
    my $id = $self->{'DIR_INFO'}->{'ID'};
    # Convert disk ID to ASCII when necessary:
    $id = D64::Disk::Image->petscii_to_ascii($id) if $convert2ascii;
    return $id;
}

=head2 get_blocks_free

Get number of blocks free:

  my $blocksFree = $d64DiskDirObj->get_blocks_free();

=cut

sub get_blocks_free {
    my $self = shift;
    $self->_check_dir_read();
    my $blocksFree = $self->{'DIR_INFO'}->{'BLOCKS_FREE'};
    return $blocksFree;
}

=head2 num_entries

Get number of directory entries:

  my $num_entries = $d64DiskDirObj->num_entries();

=cut

sub num_entries {
    my $self = shift;
    $self->_check_dir_read();
    my $entries_aref = $self->_get_dir_entries();
    my $num_entries = @{$entries_aref};
    return $num_entries;
}

=head2 get_entry

Get directory entry at the specified position (index value must be a valid position equal or greater than 0 and less than number of directory entries):

  my $entryObj = $d64DiskDirObj->get_entry($index);

Returns a valid L<D64::Disk::Dir::Entry> object upon success, and false otherwise.

=cut

sub get_entry {
    my $self = shift;
    my $index = shift;
    $self->_check_dir_read();
    my $num_entries = $self->num_entries();
    if ($index < 0 or $index >= $num_entries) {
        carp "Cannot get entry at invalid index position (disk directory contains only ${num_entries} file(s), unable to get entry at position ${index})";
        return undef;
    }
    my $entries_aref = $self->_get_dir_entries();
    my $entryObj = $entries_aref->[$index];
    return $entryObj;
}

=head2 get_file_data

Get binary file data from a directory entry at the specified position:

  my $data = $d64DiskDirObj->get_file_data($index);

Reads data from a file at the specified directory index position (index value must be a valid position equal or greater than 0 and less than number of directory entries). Returns binary file data (including its loading address) upon success, and an undefined value otherwise.

=cut

sub get_file_data {
    my $self = shift;
    my $index = shift;
    $self->_check_dir_read();
    my $entryObj = $self->get_entry($index);
    unless (defined $entryObj) {
        carp "Unable to get file data from an inexisting directory entry (validate first that ${index} file(s) really exist(s) on this disk!)";
        return undef;
    }
    my $d64DiskImageObj = $self->{'D64_DISK_IMAGE'};
    # Get filename from the specified directory index position:
    my $name = $entryObj->get_name(0);
    my $rawname = D64::Disk::Image->rawname_from_name($name);
    # Get the actual filetype:
    my $type = $entryObj->get_type();
    my $filetype = $file_type_constants{$type};
    # Open a file for reading:
    my $prg = $d64DiskImageObj->open($rawname, $filetype, F_READ);
    # Read data from file:
    my ($counter, $buffer) = $prg->read();
    # Close a file:
    $prg->close();
    return $buffer;
}

=head2 print_dir

Print out the entire directory content to any opened file handle (the standard output by default):

  $d64DiskDirObj->print_dir($fh);

=cut

sub print_dir {
    my $self = shift;
    my $fh = shift;
    $fh = *STDOUT unless defined $fh;
    $self->_check_dir_read();
    $self->_print_title($fh);
    my $num_entries = $self->num_entries();
    for (my $i = 0; $i < $num_entries; $i++) {
        my $entryObj = $self->get_entry($i);
        $entryObj->print_entry($fh);
    }
    $self->_print_blocks_free($fh);
}

sub _print_title {
    my $self = shift;
    my $fh = shift;
    # Get title converted to ASCII:
    my $title = $self->get_title(1);
    # Get disk ID converted to ASCII:
    my $id = $self->get_id(1);
    # Print title and disk ID:
    printf $fh "0 \"%-16s\" %s\n", $title, $id;
}

sub _print_blocks_free {
    my $self = shift;
    my $fh = shift;
    # Print number of blocks free:
    my $blocksFree = $self->get_blocks_free();
    printf $fh "%d blocks free\n", $blocksFree;
}

sub DESTROY {
    my $self = shift;
    $self->_release_d64_image();
}

sub _release_d64_image {
    my $self = shift;
    my $d64DiskImageObj = $self->{'D64_DISK_IMAGE'};
    delete $self->{'D64_DISK_IMAGE'};
    # Release D64 image:
    $d64DiskImageObj->free_image() if defined $d64DiskImageObj;
}

=head1 BUGS

There are no known bugs at the moment. Please report any bugs or feature requests.

=head1 EXPORT

None. No method is exported into the caller's namespace either by default or explicitly.

=head1 SEE ALSO

L<D64::Disk::Dir::Entry>, L<D64::Disk::Dir::Iterator>, L<D64::Disk::Image>

=head1 AUTHOR

Pawel Krol, E<lt>pawelkrol@cpan.orgE<gt>.

=head1 VERSION

Version 0.01 (2011-01-09)

=head1 COPYRIGHT AND LICENSE

This module is licensed under a slightly modified BSD license, the same terms as Per Olofsson's "diskimage.c" library and L<D64::Disk::Image> Perl package it is based on, license contents are repeated below.

Copyright (c) 2003-2006, Per Olofsson
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

- Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
- Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

diskimage.c website: L<http://www.paradroid.net/diskimage/>

=cut

1;


lib/D64/Disk/Dir/Iterator.pm
package D64::Disk::Dir::Iterator;

=head1 NAME

D64::Disk::Dir:Iterator - Iterating through Commodore (D64/D71/D81) disk image directory entries

=head1 SYNOPSIS

  use D64::Disk::Dir::Iterator;

  # Create an iterator with a directory object instance:
  my $iter = D64::Disk::Dir::Iterator->new($d64DiskDirObj);

  # Perlish style iterator:
  while (my $entry = $iter->getNext()) {
    # ...do something with $entry...
  }

  # C++-ish style iterator:
  for (my $iter = D64::Disk::Dir::Iterator->new($d64DiskDirObj); $iter->hasNext(); $iter->next()) {
    my $entry = $iter->current();
    # ...do something with $entry...
  }

=head1 DESCRIPTION

This package provides an iterative method of accessing individual directory entries available within D64::Disk::Dir object instance based on a simple class for iterating over Perl arrays. See the description of L<Array::Iterator> package for a complete list of available methods and iteration process examples.

=head1 METHODS

=cut

use bytes;
use strict;
use warnings;

use base qw( Exporter Array::Iterator );
our %EXPORT_TAGS = ();
$EXPORT_TAGS{'all'} = [];
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw();

our $VERSION = '0.01';

use Carp qw/carp croak verbose/;

=head2 new

Create an iterator with a D64::Disk::Dir object instance:

  my $iter = D64::Disk::Dir::Iterator->new($d64DiskDirObj);

=cut

sub new {
    my $this = shift;
    my $d64DiskDirObj = shift;
    croak "Not a D64::Disk::Dir object: \"$d64DiskDirObj\"" unless defined $d64DiskDirObj and $d64DiskDirObj->can("_get_dir_entries");
    my $entries = $d64DiskDirObj->_get_dir_entries();
    my $class = ref($this) || $this;
    my $self = $class->SUPER::new($entries);
    bless $self, $class;
    return $self;
}

=head1 BUGS

There are no known bugs at the moment. Please report any bugs or feature requests.

=head1 EXPORT

None. No method is exported into the caller's namespace either by default or explicitly.

=head1 SEE ALSO

L<Array::Iterator>, L<D64::Disk::Dir>, L<D64::Disk::Dir::Entry>, L<D64::Disk::Image>

=head1 AUTHOR

Pawel Krol, E<lt>pawelkrol@cpan.orgE<gt>.

=head1 VERSION

Version 0.01 (2011-01-09)

=head1 COPYRIGHT AND LICENSE

This module is licensed under a slightly modified BSD license, the same terms as Per Olofsson's "diskimage.c" library and L<D64::Disk::Image> Perl package it is based on, license contents are repeated below.

Copyright (c) 2003-2006, Per Olofsson
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

- Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
- Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

diskimage.c website: L<http://www.paradroid.net/diskimage/>

=cut

1;


lib/D64/Disk/Dir/Entry.pm
package D64::Disk::Dir::Entry;

=head1 NAME

D64::Disk::Dir::Entry - Handling individual Commodore (D64/D71/D81) disk image directory entries

=head1 SYNOPSIS

  use D64::Disk::Dir::Entry;

  # Create a new directory entry and initialize it with 30 bytes of binary data retrieved from a D64 disk image:
  my $entryObj = D64::Disk::Dir::Entry->new($bytes);

  # Get filename converted to ASCII string:
  my $convert2ascii = 1;
  my $name = $entryObj->get_name($convert2ascii);

  # Get various parameters describing detailed entry properties:
  my $type = $entryObj->get_type();
  my $track = $entryObj->get_track();
  my $sector = $entryObj->get_sector();

  # Print out a single line out of entire disk directory with the contents of this particular entry to the standard output:
  $entryObj->print_entry();

=head1 DESCRIPTION

This package provides a helper class for D64::Disk::Dir module, enabling user to handle individual directory entries in a higher-level object-oriented way.

=head1 METHODS

=cut

use bytes;
use strict;
use warnings;

our $VERSION = '0.01';

use Carp qw/carp croak verbose/;

use D64::Disk::Image qw(:all);

# File type names:
our @file_types = qw/del seq prg usr rel cbm dir ???/;

=head2 new

Create new C<D64::Disk::Dir::Entry> object and initialize it with 30 bytes of binary data describing each directory entry on a D64 disk image (or a physical disk):

  my $entryObj = D64::Disk::Dir::Entry->new($bytes);

The reason for initializing object not with 32 bytes of physical data but with 30 bytes instead is that two first bytes of each entry in a directory sector always should be $00 $00 as they are unused (except for the very first entry, in which case those bytes are still directory-wide, not entry-specific).

A valid C<D64::Disk::Dir::Entry> object is returned upon success, an undefined value otherwise.

=cut

sub new {
    my $this = shift;
    my $class = ref($this) || $this;
    my $self = {};
    bless $self, $class;
    my $initOK = $self->_initialize(@_);
    if ($initOK) {
        return $self;
    }
    else {
        return undef;
    }
}

sub _initialize {
    my $self = shift;
    my $bytes = shift;
    # Verify valid bytes sequence:
    unless (length $bytes == 30) {
        carp 'Initializing D64::Disk::Dir::Entry object with invalid bytes sequence (exactly 30 bytes of binary data retrieved from a physical device are required to initialize it)';
        return 0;
    }
    my $convertOK = $self->_bytes_to_data($bytes);
    return 0 unless $convertOK;
    return 1;
}

sub _bytes_to_data {
    my $self = shift;
    my $bytes = shift;
    # Get file type:
    my $file_type = ord (substr $bytes, 0x00, 0x01);
    # Get the actual filetype:
    my $type = $file_type & 7;
    # Get closed flag (not set produces "*", or "splat" files):
    my $closed = $file_type & 0x80;
    # Get locked flag (set produces ">" locked files):
    my $locked = $file_type & 0x40;
    # Get track/sector location of first sector of file:
    my $track = ord (substr $bytes, 0x01, 0x01);
    my $sector = ord (substr $bytes, 0x02, 0x01);
    # Get 16 character filename (in PETASCII, padded with $A0):
    my $rawname = substr $bytes, 0x03, 0x10;
    my $name = D64::Disk::Image->name_from_rawname($rawname);
    my ($side_track, $side_sector, $record_length) = ();
    if ($file_types[$type] eq 'rel') {
        # Get track/sector location of first side-sector block (REL file only):
        $side_track = ord (substr $bytes, 0x13, 0x01);
        $side_sector = ord (substr $bytes, 0x14, 0x01);
        # Get REL file record length (REL file only, maximum value 254):
        $record_length = ord (substr $bytes, 0x15, 0x01);
    }
    # Get file size in sectors, low/high byte order ($1C+$1D*256):
    my $size = ord (substr $bytes, 0x1D, 0x01) << 8 | ord (substr $bytes, 0x1C, 0x01);
    # Store directory entry details in a hash:
    my $dirEntry = {
        'TYPE'          => $type,
        'CLOSED'        => $closed,
        'LOCKED'        => $locked,
        'TRACK'         => $track,
        'SECTOR'        => $sector,
        'NAME'          => $name,
        'SIDE_TRACK'    => $side_track,
        'SIDE_SECTOR'   => $side_sector,
        'RECORD_LENGTH' => $record_length,
        'SIZE'          => $size,
    };
    $self->{'DETAILS'} = $dirEntry;
    return 1;
}

sub _data_to_bytes {
    my $self = shift;
    my @bytes = ();
    # Get detailed file information stored within this object instance:
    my $dirEntry = $self->{'DETAILS'};
    my $type = $dirEntry->{'TYPE'};
    my $closed = $dirEntry->{'CLOSED'};
    my $locked = $dirEntry->{'LOCKED'};
    my $track = $dirEntry->{'TRACK'};
    my $sector = $dirEntry->{'SECTOR'};
    my $name = $dirEntry->{'NAME'};
    my $side_track = $dirEntry->{'SIDE_TRACK'} || 0x00;
    my $side_sector = $dirEntry->{'SIDE_SECTOR'} || 0x00;
    my $record_length = $dirEntry->{'RECORD_LENGTH'} || 0x00;
    my $size = $dirEntry->{'SIZE'};
    # Byte $00 - File type:
    $bytes[0x00] = chr ($type | ($locked ? 0x40 : 0x00) | ($closed ? 0x80 : 0x00));
    # Byte $01 - Track location of first sector of file:
    $bytes[0x01] = chr ($track);
    # Byte $02 - Sector location of first sector of file:
    $bytes[0x02] = chr ($sector);
    # Bytes $03..$12 - 16 character filename (in PETASCII, padded with $A0):
    my $rawname = D64::Disk::Image->rawname_from_name($name);
    my $i = 0x03;
    foreach my $byte (split //, $rawname) {
        $bytes[$i++] = $byte;
    }
    # Bytes $13..$14 - Track/Sector location of first side-sector block:
    $bytes[0x13] = chr ($side_track);
    $bytes[0x14] = chr ($side_sector);
    # Byte $15 - REL file record length:
    $bytes[0x15] = chr ($record_length);
    # Bytes $16..$1B - Unused
    $bytes[0x16] = chr 0x00;
    $bytes[0x17] = chr 0x00;
    $bytes[0x18] = chr 0x00;
    $bytes[0x19] = chr 0x00;
    $bytes[0x1A] = chr 0x00;
    $bytes[0x1B] = chr 0x00;
    # Bytes $1C..$1D - File size in sectors, low/high byte order ($1C+$1D*256):
    $bytes[0x1C] = chr ($size & 0xFF);
    $bytes[0x1D] = chr (($size >> 8) & 0xFF);
    my $bytes = join '', @bytes;
    return $bytes;
}

=head2 get_type

Get the actual filetype:

  my $type = $entryObj->get_type();

Returns the actual filetype as a three-letter string, the possibilities here are: "del", "seq", "prg", "usr", "rel", "cbm", "dir", and "???".

=cut

sub get_type {
    my $self = shift;
    my $type = $self->{'DETAILS'}->{'TYPE'};
    my $file_type = $file_types[$type];
    return $file_type;
}

=head2 get_closed

Get "Closed" flag (when not set produces "*", or "splat" files):

  my $closed = $entryObj->get_closed();

Returns true when "Closed" flag is set, and false otherwise.

=cut

sub get_closed {
    my $self = shift;
    my $closed = $self->{'DETAILS'}->{'CLOSED'};
    return $closed ? 1 : 0;
}

=head2 get_locked

Get "Locked" flag (when set produces ">" locked files):

  my $locked = $entryObj->get_locked();

Returns true when "Locked" flag is set, and false otherwise.

=cut

sub get_locked {
    my $self = shift;
    my $locked = $self->{'DETAILS'}->{'LOCKED'};
    return $locked ? 1 : 0;
}

=head2 get_track

Get track location of first sector of file:

  my $track = $entryObj->get_track();

=cut

sub get_track {
    my $self = shift;
    my $track = $self->{'DETAILS'}->{'TRACK'};
    return $track;
}

=head2 get_sector

Get sector location of first sector of file:

  my $sector = $entryObj->get_sector();

=cut

sub get_sector {
    my $self = shift;
    my $sector = $self->{'DETAILS'}->{'SECTOR'};
    return $sector;
}

=head2 get_name

Get 16 character filename (in PETASCII, padded with $A0):

  my $convert2ascii = 0;
  my $name = $entryObj->get_name($convert2ascii);

Get filename converted to ASCII string:

  my $convert2ascii = 1;
  my $name = $entryObj->get_name($convert2ascii);

=cut

sub get_name {
    my $self = shift;
    my $convert2ascii = shift;
    my $name = $self->{'DETAILS'}->{'NAME'};
    $name = D64::Disk::Image->petscii_to_ascii($name) if $convert2ascii;
    return $name;
}

=head2 get_side_track

Get track location of first side-sector block (relative file only):

  my $side_track = $entryObj->get_side_track();

A track location of first side-sector block is returned upon success, an undefined value otherwise.

=cut

sub get_side_track {
    my $self = shift;
    if ($self->get_type() ne 'rel') {
        carp "Unable to get track location of first side-sector block (not a REL file!)";
        return undef;
    }
    my $side_track = $self->{'DETAILS'}->{'SIDE_TRACK'};
    return $side_track;
}

=head2 get_side_sector

Get sector location of first side-sector block (relative file only):

  my $side_sector = $entryObj->get_side_sector();

A sector location of first side-sector block is returned upon success, an undefined value otherwise.

=cut

sub get_side_sector {
    my $self = shift;
    if ($self->get_type() ne 'rel') {
        carp "Unable to get sector location of first side-sector block (not a REL file!)";
        return undef;
    }
    my $side_sector = $self->{'DETAILS'}->{'SIDE_SECTOR'};
    return $side_sector;
}

=head2 get_record_length

Get relative file record length (relative file only, maximum value 254):

  my $record_length = $entryObj->get_record_length();

A relative file record length is returned upon success, an undefined value otherwise.

=cut

sub get_record_length {
    my $self = shift;
    if ($self->get_type() ne 'rel') {
        carp "Unable to get relative file record length (not a REL file!)";
        return undef;
    }
    my $record_length = $self->{'DETAILS'}->{'RECORD_LENGTH'};
    return $record_length;
}

=head2 get_size

Get file size in sectors:

  my $size = $entryObj->get_size();

The approximate filesize in bytes is <= #sectors * 254.

=cut

sub get_size {
    my $self = shift;
    my $size = $self->{'DETAILS'}->{'SIZE'};
    return $size;
}

=head2 get_bytes

Get 30 bytes of binary data that would describe this particular directory entry on a D64 disk image (or a physical disk):

  my $bytes = $entryObj->get_bytes();

=cut

sub get_bytes {
    my $self = shift;
    my $bytes = $self->_data_to_bytes();
    return $bytes;
}

=head2 print_entry

Print entry details to any opened file handle (the standard output by default):

  $entryObj->print_entry($fh);

This method is subsequently invoked for each single entry while printing an entire directory with D64::Disk::Dir module.

=cut

sub print_entry {
    my $self = shift;
    my $fh = shift;
    $fh = *STDOUT unless defined $fh;
    # Get detailed file information stored within this object instance:
    my $type = $self->get_type();
    my $closed = $self->get_closed() ? ord ' ' : ord '*';
    my $locked = $self->get_locked() ? ord '<' : ord ' ';
    my $size = $self->get_size();
    # Get filename convert to ASCII and add quotes:
    my $name = $self->get_name(1);
    my $quotename = sprintf "\"%s\"", $name;
    # Print directory entry:
    printf $fh "%-4d  %-18s%c%s%c\n", $size, $quotename, $closed, $type, $locked;
}

=head1 BUGS

There are no known bugs at the moment. Please report any bugs or feature requests.

=head1 EXPORT

None. No method is exported into the caller's namespace either by default or explicitly.

=head1 SEE ALSO

L<D64::Disk::Dir>, L<D64::Disk::Image>

=head1 AUTHOR

Pawel Krol, E<lt>pawelkrol@cpan.orgE<gt>.

=head1 VERSION

Version 0.01 (2011-01-09)

=head1 COPYRIGHT AND LICENSE

This module is licensed under a slightly modified BSD license, the same terms as Per Olofsson's "diskimage.c" library and L<D64::Disk::Image> Perl package it is based on, license contents are repeated below.

Copyright (c) 2003-2006, Per Olofsson
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

- Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
- Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

diskimage.c website: L<http://www.paradroid.net/diskimage/>

=cut

1;
